Multiple icon definitions

mvano opened the pull request:
I think the initial patch has the general shape of what I want to do for #28.

I did find there's two ways we could go with the attribute getters on a notification, for example for the icon. To me it's not clear whether we need a single `icon` getter that reflects the selected url, or both `icon` and `icons` which reflect the exact input from options.

Previously `icon` reflected a serialization of the options `icon` input that was parsed with baseURL, not the exact original input. That means, for example, that if parsing fails, the getter returns the empty string. The implications for `NotificationIcon` are not clear to me: would invalid input such as for `type` or `sizes` need to be dropped in the same way? That might surprise developers.


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/notifications/76.html" title="Last updated on Jan 15, 2021, 7:46 AM UTC (54dd556)">Preview</a> | <a href="https://whatpr.org/notifications/76/41843b5...54dd556.html" title="Last updated on Jan 15, 2021, 7:46 AM UTC (54dd556)">Diff</a>

annevk commented on notifications.bs:
```diff
@@ -601,6 +604,39 @@ or vibrating the device again, unless the <a>renotify preference flag</a> is set
+
+  <li><p>For each <var>icon</var> in <var>icons</var> run these substeps:
+
+  <ol>
```
I know we have this indentation elsewhere, but if the `<p>` is not the sole child of an `<li>` it really should go on its own line and get indented. And this `<ol>` too.

annevk commented on notifications.bs:
```diff
@@ -601,6 +604,39 @@ or vibrating the device again, unless the <a>renotify preference flag</a> is set
+    <var>selectedIcon</var>, taking into account whether <var>icon</var>'s <code>type</code> is
+    <a lt="valid mime type">valid</a> and <a lt="resources">supported</a>, as well as the value of
+    <var>parsedSizes</var>, then set <var>selectedURL</var> to <var>iconURL</var> and set
+    <var>selectedIcon</var> to <var>icon</var>
```
Where is this paragraph from?

mvano replied:
I wrote it, but I'm not sure whether that's what you're asking. It's similar to bits of the manifest spec, but simplified, as WebIDL does a lot of the defining for us, and we don't need to describe how to parse JSON.

annevk replied:
I see, I was wondering whether HTML or manifest had a similar algorithm at some point.

annevk commented on notifications.bs:
```diff
@@ -601,6 +604,39 @@ or vibrating the device again, unless the <a>renotify preference flag</a> is set
+{{NotificationIcon}}, a <var>fallbackURLString</var> of type {{USVString}}, and
+<a lt="base url">base URL</a> <var>baseURL</var> are:
+
+<ol>
```
The steps in this list need to end in dots.

annevk commented:
I think it would be nicer if `icon`/`badge` simply reflected the selected URL.

annevk commented:
(And we'd do away with `icons`/`badges`.)

mvano commented:
Thanks Anne, I've cleaned it up a bit as discussed. Of note:
* Just return the selected URL, do not expose getters for the multiple inputs.
* As input and output are now more obviously asymmetrical (the missing getters on the output) I had to split NotificationAction into a NotificationActionOptions dictionary and a NotificationAction interface.
* Afaict using an interface with readonly attributes for the output has the happy side effect of not needing to explicitly freeze the action, so I deleted that wrinkle.

annevk commented on notifications.bs:
```diff
@@ -601,6 +604,47 @@ or vibrating the device again, unless the <a>renotify preference flag</a> is set
+<ol>
+  <li><p>Let <var>selectedURL</var> and <var>selectedIcon</var> be null.
+
+  <li><p>For each <var>icon</var> in <var>icons</var> run these substeps:
```
`<p>` needs to be on a new line.

annevk commented on notifications.bs:
```diff
@@ -601,6 +604,47 @@ or vibrating the device again, unless the <a>renotify preference flag</a> is set
+      <li>
+        <p>Let <var>iconURL</var> be the result of <a lt="url parser">parsing</a> <var>icon</var>'s
+        <code>src</code> using <var>baseURL</var>. If <var>iconURL</var> is failure, abort these
+        substeps for this <var>icon</var>.
```
Now here the `<p>` does not need to be on a new line, since it's the only child.

annevk commented:
This looks pretty good I think. Who else should we ask for review?

mvano commented:
CC @beverloo @foolip @mounirlamouri @xxyzzzq in case there's something really crazy in here. If necessary we can probably align a detail or two as needed in a followup change.

annevk commented:
I'll aim to land this Friday (maybe ping me during the day just in case) unless someone has major concerns.

foolip commented:
There's definitely some similarity here with https://github.com/whatwg/mediasession/pull/129

Feedback on that PR appreciated, and is there a good place to put the common bits? Looks to me at first glace like the whole "list of images" concept that uses the `[{src:'foo'}]` syntax can and should be shared.

foolip commented on notifications.bs:
```diff
@@ -905,19 +964,11 @@ attribute's getter must return the result of the following steps:
-         immutable. But FrozenArray reifies the dictionaries to mutable JS
-         objects accessed by reference, so we explicitly freeze them.
-         It would be better to do this with IDL primitives instead of JS - see
-         https://www.w3.org/Bugs/Public/show_bug.cgi?id=29004 -->
```
What's the conclusion of this? Looks like the issue was side-stepped by not exposing an `icons` member on `NotificationAction`, is that a permanent situation?

mvano replied:
Only exposing the selected URL seems much easier to understand. I don't see a strong use case for reflecting all the exact input of the options dictionaries, so I don't expect us to add that. Note that we already didn't expose the exact input of the icon urls, we reflected the result of resolving with baseURL.

Also, by using an interface with readonly attributes, I don't think we need to freeze the object anymore, which was a strange wrinkle.

foolip replied:
So is the resource chosen synchronously so that it can be immediately reflected through a single `icon` attribute? Doesn't that mean that the bits implementing the notification API need to know quite a bit about the UI of the platform its running on?

AFACIT, the thing that was frozen here wasn't the value returned by the attribute (which is readonly), but the `NotificationAction` objects in that array. Otherwise one can do things like `notification.actions[0].icon = 'newicon.png'` which would have no effect. Exact same problem as discussed for Media Session artwork, in other words.

mvano replied:
Good question about synchronously reflecting the selected URL... that might require an async IPC message in Chrome. I'll have to think about that one.

The general solution to avoid having to freeze dictionaries seems to be to instead return interfaces with readonly attributes. NotificationAction now is an interface that only has readonly attributes, and assigning a new value to `icon` has no effect, intentionally. The property value does not change, and internally nothing changes. No exception is thrown.

annevk commented:
Closing this due to lack of activity. If there's still interest let's discuss in #28.
